refactor!: changed layout of project

tomrtk opened the pull request:
- changed layout of project to easier build a package to deploy.
- stopped using `Poetry` for dependency management as it do not work well with nested packages.
- configured setup.cfg for building package.
- fixed issues with file paths after change of layout.
- updated README
- configured `tox` for running tests

tomrtk committed:
refactor!: changed layout of project

- changed layout of project to easier build a package to deploy.
- stopped using `Poetry` for dependency management as it do not work well with nested packages.
- configured setup.cfg for building package.
- fixed issues with file paths after change of layuout.

MindTooth commented on README.md:
````diff
@@ -1,50 +1,91 @@
 ```sh
-poetry install
+# install option 1
+virtualenv .venv
````
`python3 -m venv .venv`?

tomrtk replied:
Yes, the equivalent with python venv module.

MindTooth replied:
Any reason for not using the module?

tomrtk replied:
Both can be used. Main reason I use it, is that virtualenv is faster in creating venv's.

MindTooth commented on pyproject.toml:
```diff
@@ -78,11 +51,61 @@ verbose = true
+
+[tool.tox]
+legacy_tox_ini = """
+[tox]
```
```suggestion

[tox]
```

MindTooth reviewed (commented):
Ah.  Running of course with pre-commit edited all them files. 😅

Removing the license files because you consider these as one package?

MindTooth commented on setup.cfg:
```diff
@@ -1,3 +1,58 @@
+
+[options.entry_points]
+console_scripts =
+    nina = nina.run:main
```
Maybe rename the entrypoints in each package to be similar?

`nina.main:start`, `nina.main:serve`, etc.

tomrtk replied:
Changed the entry points to only point on the packages, it up to the package to implement how it should run in `__main__.py`.

tomrtk replied:
It did not work well. I reverted back to original structure. 

The package paths is internally in the code and entry points externally to run packages is following the same conversion, `nina`, `nina-ui`, `nina-core`, etc. 

At a later time we can rename `.py` files, function names and imports in the packages if we want, but I do not see any big benefits for doing it.

MindTooth replied:
Oh.  Maybe we could make them at share a common and good naming convention instead then?

tomrtk commented:
> Ah. Running of course with pre-commit edited all them files. 😅
> 
> Removing the license files because you consider these as one package?

Yes, that was the idea.

MindTooth commented on pyproject.toml:
```diff
@@ -78,11 +51,61 @@ verbose = true
+[tox]
+envlist = py38,py39
+isolated_build = True
+skipsdist=True
```
```suggestion
skipsdist = True
```

tomrtk committed:
fix: entry points for core, ui, etc. calling only the packages

tomrtk committed:
Update pyproject.toml

Co-authored-by: Birger J. Nordølum <contact@mindtooth.no>

tomrtk committed:
Update pyproject.toml

Co-authored-by: Birger J. Nordølum <contact@mindtooth.no>

tomrtk committed:
fix: readme mentions git-lfs

R0flcopt3r commented:
I love the new structure. It seems to play nicely with emacs too! (After I uninstalled poetry-mode..)

---

Having some issues with tox, when trying what the readme suggests for python 3.9:
```shell
tox -e py39
```
it fails with the following:
```
~/Git/fish-code new-layout
venv ❯ tox -e py39
py39 create: /home/eirik/Git/fish-code/.tox/py39
py39 installdeps: -e., pytest, coverage, requests-mock
ERROR: invocation failed (exit code 1), logfile: /home/eirik/Git/fish-code/.tox/py39/log/py39-1.log
========================================================================================================================================================== log start ===========================================================================================================================================================
ERROR: File "setup.py" not found. Directory cannot be installed in editable mode: /home/eirik/Git/fish-code
(A "pyproject.toml" file was found, but editable mode currently requires a setup.py based build.)
WARNING: You are using pip version 21.0.1; however, version 21.1.2 is available.
You should consider upgrading via the '/home/eirik/Git/fish-code/.tox/py39/bin/python -m pip install --upgrade pip' command.

=========================================================================================================================================================== log end ============================================================================================================================================================
ERROR: could not install deps [-e., pytest, coverage, requests-mock]; v = InvocationError('/home/eirik/Git/fish-code/.tox/py39/bin/python -m pip install -e. pytest coverage requests-mock', 1)
___________________________________________________________________________________________________________________________________________________________ summary ____________________________________________________________________________________________________________________________________________________________
ERROR:   py39: could not install deps [-e., pytest, coverage, requests-mock]; v = InvocationError('/home/eirik/Git/fish-code/.tox/py39/bin/python -m pip install -e. pytest coverage requests-mock', 1)
```

This happens both inside and outside the venv.

---

Most of the times building the project also fails with this:
```
~/Git/fish-code new-layout
venv ❯ pip install -e .
ERROR: File "setup.py" not found. Directory cannot be installed in editable mode: /home/eirik/Git/fish-code
(A "pyproject.toml" file was found, but editable mode currently requires a setup.py based build.)
WARNING: You are using pip version 21.0.1; however, version 21.1.2 is available.
You should consider upgrading via the '/home/eirik/Git/fish-code/venv/bin/python3 -m pip install --upgrade pip' command.
```
---

The above errors seems related? Both talking about setup.py and editable mode :shrug:

tomrtk commented:
> The above errors seems related? Both talking about setup.py and editable mode shrug

@R0flcopt3r I am not able to replicate on a freshly cloned repo. Both `tox -e py39` and `pip install -e .` works as expected in a `venv`. Have you tried to clean up the `.venv` and `.tox`?

Environment details:
```
❯ python --version --version
Python 3.9.4 (default, Apr  4 2021, 18:23:51)
[GCC 10.3.0]
❯ pip --version
pip 21.1.2 from /home/tom-ruben/tmp/test-error/.venv/lib/python3.9/site-packages/pip (python 3.9)
❯ tox --version
3.23.1 imported from /home/tom-ruben/tmp/test-error/.venv/lib/python3.9/site-packages/tox/__init__.py
```

R0flcopt3r commented:
> > The above errors seems related? Both talking about setup.py and editable mode shrug
> 
> @R0flcopt3r I am not able to replicate on a freshly cloned repo. Both `tox -e py39` and `pip install -e .` works as expected in a `venv`. Have you tried to clean up the `.venv` and `.tox`?
> 
> Environment details:
> 
> ```
> ❯ python --version --version
> Python 3.9.4 (default, Apr  4 2021, 18:23:51)
> [GCC 10.3.0]
> ❯ pip --version
> pip 21.1.2 from /home/tom-ruben/tmp/test-error/.venv/lib/python3.9/site-packages/pip (python 3.9)
> ❯ tox --version
> 3.23.1 imported from /home/tom-ruben/tmp/test-error/.venv/lib/python3.9/site-packages/tox/__init__.py
> ```

Fresh venv, and it still fails:

```
~/Git/fish-code new-layout
❯ make clean-all
find . -name ".venv" -type d -exec rm -f -r {} +
find . -name "__pycache__" -type d -exec rm -f -r {} +
find . -name "*.egg-info" -type d -exec rm -f -r {} +
find . -name "instance" -type d -exec rm -f -r {} +
find . -name ".tox" -type d -exec rm -f -r {} +
find . -name ".pytest_cache" -type d -exec rm -f -r {} +
find . -name ".coverage" -type f -delete
rm -f data.db
rm -f -r dist

~/Git/fish-code new-layout
❯ virtualenv .venv
created virtual environment CPython3.9.5.final.0-64 in 174ms
  creator CPython3Posix(dest=/home/eirik/Git/fish-code/.venv, clear=False, no_vcs_ignore=False, global=False)
  seeder FromAppData(extra_search_dir=/usr/share/python-wheels,download=False, pip=bundle, setuptools=bundle, wheel=bundle, via=copy, app_data_dir=/home/eirik/.local/share/virtualenv)
    added seed packages: pip==21.0.1, setuptools==53.0.0, wheel==0.36.2
  activators BashActivator,CShellActivator,FishActivator,PowerShellActivator,PythonActivator,XonshActivator

~/Git/fish-code new-layout
❯ . .venv/bin/activate

~/Git/fish-code new-layout
.venv ❯ pip install -e .
ERROR: File "setup.py" not found. Directory cannot be installed in editable mode: /home/eirik/Git/fish-code
(A "pyproject.toml" file was found, but editable mode currently requires a setup.py based build.)
WARNING: You are using pip version 21.0.1; however, version 21.1.2 is available.
You should consider upgrading via the '/home/eirik/Git/fish-code/venv/bin/python3 -m pip install --upgrade pip' command.
```

tomrtk commented:
@R0flcopt3r 
> WARNING: You are using pip version 21.0.1; however, version 21.1.2 is available.
You should consider upgrading via the '/home/eirik/Git/fish-code/venv/bin/python3 -m pip install --upgrade pip' command.

R0flcopt3r commented:
> @R0flcopt3r
> 
> > WARNING: You are using pip version 21.0.1; however, version 21.1.2 is available.
> > You should consider upgrading via the '/home/eirik/Git/fish-code/venv/bin/python3 -m pip install --upgrade pip' command.

that did fix it.

How come that shows up even though I have updated pip globally?

tomrtk committed:
fix: path in testfiles

tomrtk committed:
chore: pip need to be of v21.1 or higher

tomrtk committed:
fix: revert setup.cfg entrypoints

MindTooth commented:
> > @R0flcopt3r
> > > WARNING: You are using pip version 21.0.1; however, version 21.1.2 is available.
> > > You should consider upgrading via the '/home/eirik/Git/fish-code/venv/bin/python3 -m pip install --upgrade pip' command.
> 
> that did fix it.
> 
> How come that shows up even though I have updated pip globally?

 Both were envs I guess.  Manually sourced or through `tox`.

tomrtk commented:
> > @R0flcopt3r
> > > WARNING: You are using pip version 21.0.1; however, version 21.1.2 is available.
> > > You should consider upgrading via the '/home/eirik/Git/fish-code/venv/bin/python3 -m pip install --upgrade pip' command.
> 
> that did fix it.
> 
> How come that shows up even though I have updated pip globally?

Believe it is because it is a fresh install of pip in `venv`, and `pip` will be the version shipped with that python version.

R0flcopt3r committed:
build: adds new targets to makefile

Makes the makefile the one stop shop for anything related to building
this code. It will set up virtual environment, install dependencies and
let you easily run the software.

This removes any fear of polluting your system with pip packages by
mistake. It also allows you to rename the virtual environment to
anything that suits you. It is not recommended to rename the virtual
environment to `venv`. I have tested and it works, but produces
warnings.

updates .gitignore to not track files generated by make.

MindTooth commented on README.md:
```diff
@@ -1,50 +1,93 @@
+git clone <url> nina
+cd nina
+# make virtual environment
+virtualenv .venv
```
Can add `--download` to make sure that the latest version of pip is installed.

Ref. https://virtualenv.pypa.io/en/latest/cli_interface.html#download

R0flcopt3r replied:
yepp, that seems to fix the issue I were having.

MindTooth commented on README.md:
```diff
@@ -1,50 +1,93 @@
+. .venv/bin/activate
+# get dependencies
+pip install -e .
+pip install -r requirements-dev.txt
```
Don't both commands provide the same dependencies?  Could one be forfeited for the other?

tomrtk replied:
No.
 `-e .` installs the package in editable mode. `requirements` only installs dependencies.

MindTooth commented:
And the one error that I often encounter:

```
  ERROR: Command errored out with exit status 1:
   command: /Users/mt/School/fish-code/.venv/bin/python -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/private/var/folders/xf/_h8ss72n1v51lyp86t1qxp1m0000gn/T/pip-install-ycoqy7za/lap_0e318d8ff7a04b19b8a9d57cd1b8377d/setup.py'"'"'; __file__='"'"'/private/var/folders/xf/_h8ss72n1v51lyp86t1qxp1m0000gn/T/pip-install-ycoqy7za/lap_0e318d8ff7a04b19b8a9d57cd1b8377d/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /private/var/folders/xf/_h8ss72n1v51lyp86t1qxp1m0000gn/T/pip-wheel-1sy1llvl
       cwd: /private/var/folders/xf/_h8ss72n1v51lyp86t1qxp1m0000gn/T/pip-install-ycoqy7za/lap_0e318d8ff7a04b19b8a9d57cd1b8377d/
  Complete output (12 lines):
  Partial import of lap during the build process.
  Traceback (most recent call last):
    File "/private/var/folders/xf/_h8ss72n1v51lyp86t1qxp1m0000gn/T/pip-install-ycoqy7za/lap_0e318d8ff7a04b19b8a9d57cd1b8377d/setup.py", line 127, in get_numpy_status
      import numpy
  ModuleNotFoundError: No module named 'numpy'
  Traceback (most recent call last):
    File "<string>", line 1, in <module>
    File "/private/var/folders/xf/_h8ss72n1v51lyp86t1qxp1m0000gn/T/pip-install-ycoqy7za/lap_0e318d8ff7a04b19b8a9d57cd1b8377d/setup.py", line 236, in <module>
      setup_package()
    File "/private/var/folders/xf/_h8ss72n1v51lyp86t1qxp1m0000gn/T/pip-install-ycoqy7za/lap_0e318d8ff7a04b19b8a9d57cd1b8377d/setup.py", line 220, in setup_package
      raise ImportError('lap requires numpy, '
  ImportError: lap requires numpy, please "pip install numpy".
  ----------------------------------------
  ERROR: Failed building wheel for lap
```

The lap package is quite old and doesn't use new logic in how to calculate deps.  Have you guys come across the issue?

---

Maybe we could change from waitress to uvicorn for Flask too?  Don't they both provide the same?

R0flcopt3r commented:
> And the one error that I often encounter:
> 
> ```
>   ERROR: Command errored out with exit status 1:
>    command: /Users/mt/School/fish-code/.venv/bin/python -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/private/var/folders/xf/_h8ss72n1v51lyp86t1qxp1m0000gn/T/pip-install-ycoqy7za/lap_0e318d8ff7a04b19b8a9d57cd1b8377d/setup.py'"'"'; __file__='"'"'/private/var/folders/xf/_h8ss72n1v51lyp86t1qxp1m0000gn/T/pip-install-ycoqy7za/lap_0e318d8ff7a04b19b8a9d57cd1b8377d/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /private/var/folders/xf/_h8ss72n1v51lyp86t1qxp1m0000gn/T/pip-wheel-1sy1llvl
>        cwd: /private/var/folders/xf/_h8ss72n1v51lyp86t1qxp1m0000gn/T/pip-install-ycoqy7za/lap_0e318d8ff7a04b19b8a9d57cd1b8377d/
>   Complete output (12 lines):
>   Partial import of lap during the build process.
>   Traceback (most recent call last):
>     File "/private/var/folders/xf/_h8ss72n1v51lyp86t1qxp1m0000gn/T/pip-install-ycoqy7za/lap_0e318d8ff7a04b19b8a9d57cd1b8377d/setup.py", line 127, in get_numpy_status
>       import numpy
>   ModuleNotFoundError: No module named 'numpy'
>   Traceback (most recent call last):
>     File "<string>", line 1, in <module>
>     File "/private/var/folders/xf/_h8ss72n1v51lyp86t1qxp1m0000gn/T/pip-install-ycoqy7za/lap_0e318d8ff7a04b19b8a9d57cd1b8377d/setup.py", line 236, in <module>
>       setup_package()
>     File "/private/var/folders/xf/_h8ss72n1v51lyp86t1qxp1m0000gn/T/pip-install-ycoqy7za/lap_0e318d8ff7a04b19b8a9d57cd1b8377d/setup.py", line 220, in setup_package
>       raise ImportError('lap requires numpy, '
>   ImportError: lap requires numpy, please "pip install numpy".
>   ----------------------------------------
>   ERROR: Failed building wheel for lap
> ```
> 
> The lap package is quite old and doesn't use new logic in how to calculate deps. Have you guys come across the issue?
> 
> Maybe we could change from waitress to uvicorn for Flask too? Don't they both provide the same?

This happens for me too. Doesn't seem to matter. Numpy is getting installed and everything seems to be functioning normally. I tried moving numpy further up, didn't solve it.

MindTooth commented on README.md:
```diff
@@ -1,50 +1,93 @@
+# get dependencies
+pip install -e .
+pip install -r requirements-dev.txt
+# run application with Flask dev-server
```
Maybe check for `FLASK_ENV=development` instead of providing `--dev`.  Now it still runs in production mode even though `--dev` is passed.

MindTooth replied:
```
usage: __main__.py [-h] [--host HOST] [--port PORT] [--debug] [--test]
__main__.py: error: unrecognized arguments: --dev
usage: __main__.py [-h] [--debug] [--host HOST] [--port PORT] [--test]
__main__.py: error: unrecognized arguments: --dev
```

tomrtk replied:
```
❯ ./.venv/bin/python -m nina
INFO:core.main:Core started
INFO:core.main:Creating database engine
INFO:tracing.main:Tracing started
INFO:detection.main:Detection started
INFO:waitress:Serving on http://0.0.0.0:5000
INFO:core.main:Creating database schema
INFO:     Started server process [28316]
INFO:     Started server process [28317]
..,
```
If I pass `dev` flag it runs with flask dev server.
```
❯ ./.venv/bin/python -m nina --dev
INFO:core.main:Core started
INFO:core.main:Creating database engine
usage: __main__.py [-h] [--host HOST] [--port PORT] [--debug] [--test]
__main__.py: error: unrecognized arguments: --dev
usage: __main__.py [-h] [--debug] [--host HOST] [--port PORT] [--test]
__main__.py: error: unrecognized arguments: --dev
 * Serving Flask app "ui.main" (lazy loading)
 * Environment: production
   WARNING: This is a development server. Do not use it in a production deployment.
   Use a production WSGI server instead.
 * Debug mode: on
INFO:core.main:Creating database schema
INFO:core.repository.orm:Starting mappers
```

tomrtk replied:
> Maybe check for FLASK_ENV=development instead of providing --dev. 

Feel free to change it if you feel it is better.

MindTooth commented on Makefile:
```diff
@@ -1,18 +1,59 @@
 
-.PHONY: help clean-all clean-poetry-lock clean-venv
+$(venv)/bin/nina: $(src_files) $(venv)
+	@. $(activate) &&\
```
```suggestion
	@. $(activate) && \
```

MindTooth commented on Makefile:
```diff
@@ -1,18 +1,59 @@
-	find . -name ".venv" -type d -exec $(RM) -r {} +
+	find . -name "$(venv)" -type d -exec $(RM) -r {} +
+	find . -name ".make.deps" -type f -exec $(RM) -f {} +
+
```
```suggestion
```

MindTooth commented on Makefile:
```diff
@@ -1,18 +1,59 @@
+	@. $(activate) &&\
+	python -m nina
+
+$(venv):
```
How does this work as this is not a PHONY target?

R0flcopt3r replied:
Works as expected. If the venv exist, it won't do anything.

MindTooth commented on Makefile:
```diff
@@ -1,18 +1,59 @@
+	touch $@
 
-.PHONY: help clean-all clean-poetry-lock clean-venv
+$(venv)/bin/nina: $(src_files) $(venv)
```
And this for a PHONY target?

R0flcopt3r replied:
Why? I want it to not run the target if it's up to date.

MindTooth commented on Makefile:
```diff
@@ -1,18 +1,59 @@
+activate := $(venv)/bin/activate
+src_files := $(shell find src -type f -iname '*.py')
+
+.PHONY: help clean-all clean-venv run shell
```
```suggestion
.PHONY: help clean-all clean-venv run shell deps venv nina
```

MindTooth commented on Makefile:
```diff
@@ -1,18 +1,59 @@
+nina: $(venv)/bin/nina
+
+run: nina
+	@. $(activate) &&\
```
```suggestion
	@. $(activate) && \
```
